README.md
# [DateTime][docsrs]: High-Level Date and Time for Rust

[![DateTime on crates.io][cratesio-image]][cratesio]
[![DateTime on docs.rs][docsrs-image]][docsrs]
[![License][apache-image]][apache]
[![Downloads on Crates.io][cratesio-downloads-image]][cratesio]

[![Build Status][github-actions-image]][github-actions]
[![FOSSA Status][fossa-image]][fossa]
[![Coverage Status][coveralls-image]][coveralls]
[![CII Best Practices Status][cii-best-practices-image]][cii-best-practices]

[![Average time to resolve an issue][isitmaintained-resolution-image]][isitmaintained]
[![Percentage of issues still open][isitmaintained-open-image]][isitmaintained]

[github-actions-image]: https://img.shields.io/endpoint.svg?url=https%3A%2F%2Factions-badge.atrox.dev%2Fnotquiteamonad%2Fdate_time%2Fbadge&style=flat

[github-actions]: https://actions-badge.atrox.dev/notquiteamonad/date_time/goto

[cratesio-image]: https://img.shields.io/crates/v/date_time.svg

[cratesio]: https://crates.io/crates/date_time

[docsrs-image]: https://docs.rs/date_time/badge.svg

[docsrs]: https://docs.rs/date_time

[apache-image]: https://img.shields.io/badge/License-Apache%202.0-blue.svg

[apache]: https://opensource.org/licenses/Apache-2.0

[cratesio-downloads-image]: https://img.shields.io/crates/d/date_time

[fossa]: https://app.fossa.io/projects/git%2Bgithub.com%2Fnotquiteamonad%2Fdate_time?ref=badge_shield

[fossa-image]: https://app.fossa.io/api/projects/git%2Bgithub.com%2Fnotquiteamonad%2Fdate_time.svg?type=shield

[isitmaintained]: http://isitmaintained.com/project/notquiteamonad/date_time

[isitmaintained-resolution-image]: http://isitmaintained.com/badge/resolution/notquiteamonad/date_time.svg

[isitmaintained-open-image]: http://isitmaintained.com/badge/open/notquiteamonad/date_time.svg

[coveralls]: https://coveralls.io/github/notquiteamonad/date_time

[coveralls-image]: https://coveralls.io/repos/github/notquiteamonad/date_time/badge.svg

[cii-best-practices]: https://bestpractices.coreinfrastructure.org/projects/3127

[cii-best-practices-image]: https://bestpractices.coreinfrastructure.org/projects/3127/badge

Date_Time is a high-level rust library for use in situations where
precision beyond seconds is not necessary.

It handles serialisable dates and times from 01 Jan 0000 at 00:00:00 to
Dec 31 9999 at 23:59:59.

**Note:** This README may contain documentation of features which have not been released yet. To ensure
it is accurate for a specific version, use the tag selector and select the version you're using.

[Changelog](https://github.com/notquiteamonad/date_time/blob/master/CHANGELOG.md) - **Note Version 2.0.0 includes a breaking change.**

[Code of Conduct](https://github.com/notquiteamonad/date_time/blob/master/CODE_OF_CONDUCT.md)

## Usage

Put this in your `Cargo.toml`:

```toml
[dependencies]
date_time = "2"
```

## Overview

All of the types in this library implement `Debug`, `Copy`, and `Clone` traits.

##### Naming

This library was originally a port from a closed-source Kotlin library with
similar functionality. As such, each of the types are suffixed with "Tuple".

Type aliases exist without the Tuple suffixes from version 1.2.1 onwards.

There is a new replacement of the Kotlin library available
[here](https://github.com/notquiteamonad/date-time-kt).

##### Serde Support

Types exported by this library have `Serialize` and `Deserialize` implementations
behind the `serde_support` feature flag.

### Times

#### TimeTuple

Times can be generated using the `timetuple::TimeTuple` type.

Times must either be instantiated using `TimeTuple::new()` which takes hour, minute, and second parameters or `TimeTuple::from_seconds()`, which just takes a total number of seconds. These are then converted into seconds and split apart again to create a tuple between 00:00:00 and 23:59:59.

TimeTuple implements the following traits and is therefore fully comparable with other TimeTuples.

-   `PartialOrd`
-   `Ord`
-   `PartialEq`
-   `Eq`

It can also be added to and subtracted from another TimeTuple, but the user must be aware that this will loop around midnight.

For example:

-   `TimeTuple::new(22, 0, 0) + TimeTuple::new(1, 0, 0)` will produce `TimeTuple { h: 23, m: 0, s:0 }`
-   `TimeTuple::new(22, 0, 0) + TimeTuple::new(3, 0, 0)` will produce `TimeTuple { h: 1, m: 0, s:0 }`

##### Serialisation

`TimeTuple` can be serialised using `to_string()` (generated from Display trait) and `to_hhmm_string()`.

For 8:30:30 AM, the former will produce `"08:30:30"` and the latter will produce `"08:30"`.

A `TimeTuple` can be instantiated by calling `TimeTuple::from_str()` with a string in the format of `hh:mm:ss`.

##### Mutation

The following methods exist to manipulate an existing `TimeTuple`:

-   `add_seconds()`
-   `subtract_seconds()`
-   `add_minutes()`
-   `subtract_minutes()`
-   `add_hours()`
-   `subtract_hours()`

Each takes a single argument of the number to add/subtract. These methods all wrap such that the resulting time is a valid time between `00:00:00` and `23:59:59`.

#### Duration

A second time type, `Duration` exists for cases where a duration should be stored in hours, minutes and seconds. This is similar to the `TimeTuple` type but allows hours to be greater than 24.

The difference between any two `DateTime`s can be calculated using `Duration::between()`.

### Dates

Dates can be generated using the `datetuple::DateTuple` and `monthtuple::MonthTuple` types. The `MonthTuple` type is similar to `DateTuple` but doesn't include a day of the month.

#### DateTuple

`DateTuple` wraps a year, month, and day of month in a struct.

A `DateTuple` can be created using `DateTuple::new()`, passing a year between 0 and 9999 and a month and date which are valid for that year. Feb 29 can be created if the year is a leap year.

A `DateTuple` can be converted to and from a number of days as a `u32` using the `to_days()` and `from_days()` methods. The number of days referenced is the number of days between 0000-01-01 and the `DateTuple`'s value inclusive.

`DateTuple` is fully comparable with another `DateTuple` and implements `PartialOrd`, `Ord`, `PartialEq`, and `Eq`.

##### Serialisation

`DateTuple` can be serialised using `to_string()` (generated from Display trait) and `to_readable_string()`.

For 23rd January 2002, the former will produce `"2002-01-23"` and the latter will produce `"23 Jan 2002"`.

A `DateTuple` can be instantiated by calling `DateTuple::from_str()` with a string in the format of `yyyy-mm-dd`.

If listing multiple `DateTuple` objects in a human readable format, you may wish to pad them with a space to the left to ensure alignment. This can be done with the format specifier `{:>11}` in a call such as `format!()`.

##### Mutation

The following methods exist to manipulate an existing `DateTuple`:

-   `add_days()`
-   `subtract_days()`
-   `add_months()`
-   `subtract_months()`
-   `add_years()`
-   `subtract_years()`

Each takes a single argument of the number to add/subtract. These methods will always return a valid date. If the date were to fall after the end of a month, such as after adding one year to Feb 29 on a leap year, the last valid date in the month will be returned.

The following two methods consume a `DateTuple` and return another:

-   `next_date()`
-   `previous_date()`

They work similarly to `next_month()` and `previous_month()` described below.

#### MonthTuple

`MonthTuple` is identical to `DateTuple` but without a day of the month.

It can be instantiated using `MonthTuple::new()`, passing a year between 0000 and 9999 and a month between 1 and 12.

`MonthTuple` is fully comparable with another `MonthTuple` and implements `PartialOrd`, `Ord`, `PartialEq`, and `Eq`.

`MonthTuple` also implements `From<DateTuple>` so a `DateTuple` can be converted to a `MonthTuple` using `MonthTuple::from(date_tuple: DateTuple)`.

##### Mutation

The following methods exist to manipulate an existing `MonthTuple`:

-   `add_months()`
-   `subtract_months()`
-   `add_years()`
-   `subtract_years()`

Each takes a single argument of the number to add/subtract.

###### `next_month` and `previous_month`

`MonthTuple` provides two methods: `next_month` and `previous_month` which consume the `MonthTuple` and return the `MonthTuple` which chronologically follows or precedes it.

These will continue to return the maximum and minimum values of Jan 0000 and Dec 9999 if they are reached.

These methods consume the existing `MonthTuple`.

##### Serialisation

`MonthTuple` can be serialised using `to_string()` (generated from Display trait) and `to_readable_string()`.

For January 2002, the former will produce `"2002-01"` and the latter will produce `"Jan 2002"`.

A `MonthTuple` can be instantiated by calling `MonthTuple::from_str()` with a string in the format of `yyyy-mm`.

### DateTime

The `date_time_tuple::DateTimeTuple` type wraps a `DateTuple` and a `TimeTuple`.

Like the other modules in this library, it is fully comparable with other `DateTimeTuple` structs.

The difference between two `DateTime`s can be calculated using `Duration::between()`.

##### Serialisation

`DateTimeTuple` can be serialised using `to_string()` (generated from Display trait) and `to_readable_string()`.

For 23rd January 2002 at 08:30:30 AM, the former will produce `"2002-01-23@08:30:30"` and the latter will produce `"23 Jan 2002 08:30:30"`.

A `DateTimeTuple` can be instantiated by calling `DateTimeTuple::from_str()` with a string in the format of `yyyy-mm-dd@hh:mm:ss`.

## Limitations

This library was designed for high-level implementations of dates in which precision is not necessary.

For a more precise wrapper of dates, try a crate such as [chrono](https://crates.io/crates/chrono).

-   This library is only designed for use when dates need only to be precise to the level of seconds.
-   This library is timezone-agnostic; it doesn't deal with any difference between time zones.
-   Only datetimes between `01 Jan 0000 00:00:00` and `31 Dec 9999 23:59:59` are supported.

## Vulnerabilities

Vulnerabilities should be reported through GitHub issues (or confidentially by email to 44178347+notquiteamonad@users.noreply.github.com).

Dependabot is being used by this repository to ensure dependencies are kept up-to-date, meaning security fixes will be applied as quickly as possible once vulnerabilities are patched in dependencies.


CHANGELOG.md
# Changelog

## Version 2.2.0

- Derived `Hash` on all types (#61, thanks @flauntingspade4)
- Derived `Serialize` and `Deserialise` behind the `serde_support` flag (also #61)

## Version 2.1.1

- Improved performance of functions using regex (thanks, @tuube)
- Updated maintainer name to @notquiteamonad

## Version 2.1.0

- Added `between` method to get the `Duration` between two `DateTime`s.
- Deprecated `to_hours_and_minutes_string` on `Duration` in favour of `to_hhmm_string` for consistency with `TimeTuple`.

## Version 2.0.1

- Minor documentation improvements

## Version 2.0.0 - _Includes Breaking Changes_

-   **BREAKING CHANGE:** Months in `MonthTuple` and `DateTuple` are now one-based rather than zero-based (i.e. 1 represents January). If you have any serialised data from this library which is to be reingested in code, all months will need to be incremented by 1 before use with version 2.0.0.
-   Added `min_value()` and `max_value()` to `DateTuple`.
-   Added `to_days()` and `from_days()` to `DateTuple`.

## Version 1.5.0

-   Changed `MonthTuple` and `DateTuple`'s serialisation to conform to [ISO 8601](https://www.iso.org/iso-8601-date-and-time-format.html). Legacy formats can still be parsed via the `FromStr` implementation.

## Version 1.4.4

-   Add `to_minutes` methods for `TimeTuple` and `Duration`
-   Maximised test coverage

## Versions 1.4.2 and 1.4.3

-   Improved documentation

## Version 1.4.1

-   Updated documentation

## Version 1.4.0

-   Added the `Duration` type to hold a time longer than 24 hours.

## Version 1.3.0

-   Added methods to get the next and previous date/month from a `DateTuple`/`MonthTuple`.
-   All of the types now have methods which mutate them by either adding or subtracting discrete values of one of their components.
-   `MonthTuple::this_month()`, `DateTuple::today()`, and `TimeTuple::now()` provide access to values produced from the current time as provided by `std::time::SystemTime::now()`.
-   A `TimeTuple` can now be produced from a number of seconds using `TimeTuple::from_seconds()`.

## Version 1.2.1

-   Added type aliases of `Date`, `Month`, and `Time` for `DateTuple`, `MonthTuple`, and `TimeTuple` respectively.

## Version 1.2.0

-   First public release


tests/month_tuple_tests.rs
extern crate date_time;

use date_time::month_tuple::MonthTuple;

#[test]
fn test_component_too_large() {
    assert!(MonthTuple::new(2000, 12).is_ok());
    assert!(MonthTuple::new(2000, 13).is_err());
    assert!(MonthTuple::new(10000, 5).is_err());
}

#[test]
fn test_this_month_does_not_panic() {
    MonthTuple::this_month();
}

#[test]
fn test_to_readable_string() {
    let tuple = MonthTuple::new(2000, 5).unwrap();
    assert_eq!(String::from("May 2000"), tuple.to_readable_string());
}

#[test]
fn test_to_string() {
    let tuple = MonthTuple::new(2000, 5).unwrap();
    assert_eq!(String::from("2000-05"), tuple.to_string());
}

#[test]
fn test_equals() {
    let tuple1 = MonthTuple::new(2000, 5).unwrap();
    let tuple2 = MonthTuple::new(2000, 5).unwrap();
    assert_eq!(tuple1, tuple2);
}

#[test]
fn test_comparisons() {
    let tuple1 = MonthTuple::new(2000, 5).unwrap();
    let tuple2 = MonthTuple::new(2000, 5).unwrap();
    let tuple3 = MonthTuple::new(2000, 6).unwrap();
    let tuple4 = MonthTuple::new(2001, 1).unwrap();
    assert!(tuple1 <= tuple2);
    assert!(!(tuple1 < tuple2));
    assert!(tuple1 >= tuple2);
    assert!(tuple1 < tuple3);
    assert!(tuple3 < tuple4);
    assert!(tuple4 > tuple2);
}

#[test]
fn test_from_string() {
    let tuple = MonthTuple::new(2000, 5).unwrap();
    assert_eq!(tuple, str::parse("2000-05").unwrap());
    assert_eq!(tuple, str::parse("200005").unwrap());
    assert!(str::parse::<MonthTuple>("2000-15").is_err());
    assert!(str::parse::<MonthTuple>("200015").is_err());
    assert!(str::parse::<MonthTuple>("200O05").is_err());
}

#[test]
fn test_add_months() {
    let mut tuple1 = MonthTuple::new(2000, 6).unwrap();
    let tuple1_orig = MonthTuple::new(2000, 6).unwrap();
    let mut tuple2 = MonthTuple::new(2000, 12).unwrap();
    let tuple2_orig = MonthTuple::new(2000, 12).unwrap();
    tuple1.add_months(1);
    assert_eq!(tuple1, tuple1_orig.next_month());
    tuple2.add_months(2);
    assert_eq!(tuple2, tuple2_orig.next_month().next_month());
}

#[test]
fn test_subtract_months() {
    let mut tuple1 = MonthTuple::new(2000, 6).unwrap();
    let tuple1_orig = MonthTuple::new(2000, 6).unwrap();
    let mut tuple2 = MonthTuple::new(2000, 12).unwrap();
    let tuple2_orig = MonthTuple::new(2000, 12).unwrap();
    tuple1.subtract_months(1);
    assert_eq!(tuple1, tuple1_orig.previous_month());
    tuple2.subtract_months(2);
    assert_eq!(tuple2, tuple2_orig.previous_month().previous_month());
}

#[test]
fn test_add_years() {
    let mut tuple1 = MonthTuple::new(2000, 6).unwrap();
    let mut tuple2 = MonthTuple::new(9998, 6).unwrap();
    tuple1.add_years(2);
    assert_eq!(2002, tuple1.get_year());
    tuple2.add_years(2);
    assert_eq!(9999, tuple2.get_year());
}

#[test]
fn test_subtract_years() {
    let mut tuple1 = MonthTuple::new(2000, 6).unwrap();
    let mut tuple2 = MonthTuple::new(1, 6).unwrap();
    tuple1.subtract_years(2);
    assert_eq!(1998, tuple1.get_year());
    tuple2.subtract_years(2);
    assert_eq!(0, tuple2.get_year());
}


src/time_tuple.rs
use crate::date_time_tuple::DateTimeTuple;
use date_utils;
use regex::Regex;
use std::cmp::Ordering;
use std::convert::From;
use std::fmt;
use std::ops::{Add, AddAssign, Sub, SubAssign};
use std::str::FromStr;

pub type Time = TimeTuple;
pub type TimeOfDay = TimeTuple;

/// A wrapper for a particular time of day.
///
/// Precise to second-level.
///
/// **NOTE:** This cannot be 24 hours or greater - see `TimeTuple::new()` for more details.
///
/// The wrapping described in `TimeTuple::new()` also applies when adding and subtracting times.
#[cfg_attr(
    feature = "serde_support",
    derive(serde::Serialize, serde::Deserialize)
)]
#[derive(PartialEq, Eq, Debug, Copy, Clone, Hash)]
pub struct TimeTuple {
    h: u8,
    m: u8,
    s: u8,
}

impl TimeTuple {
    /// Produces a new TimeTuple.
    ///
    /// Times of 24 hours or greater and negative times
    /// will wrap around 24 hours to always produce a positive time.
    ///
    /// The value is calculated from total number of seconds so a time
    /// with a minute value of 90 would add an hour to the resulting tuple
    /// and set the minutes to 30, for example.
    pub fn new(h: i32, m: i32, s: i32) -> TimeTuple {
        let mut total_seconds = s + 60 * m + 3600 * h;
        while total_seconds < 0 {
            total_seconds += 86400;
        }
        TimeTuple::from_seconds(total_seconds as u64)
    }

    /// Same as `TimeTuple::new()` but takes the total number of seconds
    /// as its argument and calculates the hours, minutes, and seconds
    /// from that.
    pub fn from_seconds(mut total_seconds: u64) -> TimeTuple {
        while total_seconds >= 86400 {
            total_seconds -= 86400;
        }
        let h = total_seconds / 3600;
        total_seconds -= h * 3600;
        let m = total_seconds / 60;
        total_seconds -= m * 60;
        TimeTuple {
            h: h as u8,
            m: m as u8,
            s: total_seconds as u8,
        }
    }

    /// Returns a `TimeTuple` of the current time as `std::time::SystemTime` provides it.
    pub fn now() -> TimeTuple {
        date_utils::now_as_timetuple()
    }

    pub fn get_hours(self) -> u8 {
        self.h
    }

    pub fn get_minutes(self) -> u8 {
        self.m
    }

    pub fn get_seconds(self) -> u8 {
        self.s
    }

    /// Produces a string such as 08:30 for 8 hours and 30 minutes.
    ///
    /// Ignores seconds.
    pub fn to_hhmm_string(self) -> String {
        format!("{:02}:{:02}", self.h, self.m)
    }

    /// Gets the total number of seconds in the tuple.
    pub fn to_seconds(self) -> u32 {
        3600 * u32::from(self.h) + 60 * u32::from(self.m) + u32::from(self.s)
    }

    /// Gets the total number of minutes in the tuple, ignoring seconds.
    pub fn to_minutes(self) -> u32 {
        60 * u32::from(self.h) + u32::from(self.m)
    }

    /// Adds a number of seconds to the TimeTuple,
    /// wrapping the same way `TimeTuple::new()` does.
    pub fn add_seconds(&mut self, seconds: i32) {
        let new_seconds = i32::from(self.s) + seconds;
        *self = TimeTuple::new(i32::from(self.h), i32::from(self.m), new_seconds);
    }

    /// Subtracts a number of seconds from the TimeTuple,
    /// wrapping the same way `TimeTuple::new()` does.
    pub fn subtract_seconds(&mut self, seconds: i32) {
        let new_seconds = i32::from(self.s) - seconds;
        *self = TimeTuple::new(i32::from(self.h), i32::from(self.m), new_seconds);
    }

    /// Adds a number of minutes to the TimeTuple,
    /// wrapping the same way `TimeTuple::new()` does.
    pub fn add_minutes(&mut self, minutes: i32) {
        let new_minutes = i32::from(self.m) + minutes;
        *self = TimeTuple::new(i32::from(self.h), new_minutes, i32::from(self.s));
    }

    /// Subtracts a number of minutes from the TimeTuple,
    /// wrapping the same way `TimeTuple::new()` does.
    pub fn subtract_minutes(&mut self, minutes: i32) {
        let new_minutes = i32::from(self.m) - minutes;
        *self = TimeTuple::new(i32::from(self.h), new_minutes, i32::from(self.s));
    }

    /// Adds a number of hours to the TimeTuple,
    /// wrapping the same way `TimeTuple::new()` does.
    pub fn add_hours(&mut self, hours: i32) {
        let new_hours = i32::from(self.h) + hours;
        *self = TimeTuple::new(new_hours, i32::from(self.m), i32::from(self.s));
    }

    /// Subtracts a number of hours from the TimeTuple,
    /// wrapping the same way `TimeTuple::new()` does.
    pub fn subtract_hours(&mut self, hours: i32) {
        let new_hours = i32::from(self.h) - hours;
        *self = TimeTuple::new(new_hours, i32::from(self.m), i32::from(self.s));
    }
}

impl fmt::Display for TimeTuple {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{:02}:{:02}:{:02}", self.h, self.m, self.s)
    }
}

impl FromStr for TimeTuple {
    type Err = String;

    fn from_str(s: &str) -> Result<TimeTuple, Self::Err> {
        lazy_static! {
            static ref VALID_FORMAT: Regex = Regex::new(r"^\d{2}:\d{2}:\d{2}$").unwrap();
        }

        if !VALID_FORMAT.is_match(s) {
            Err(format!(
                "Invalid str formatting of TimeTuple: {}\nExpects a string formatted like 08:30:05",
                s
            ))
        } else {
            let mut parts = s.split(':');
            Ok(TimeTuple::new(
                i32::from_str(parts.next().unwrap()).unwrap(),
                i32::from_str(parts.next().unwrap()).unwrap(),
                i32::from_str(parts.next().unwrap()).unwrap(),
            ))
        }
    }
}

impl PartialOrd for TimeTuple {
    fn partial_cmp(&self, other: &TimeTuple) -> Option<Ordering> {
        self.to_seconds().partial_cmp(&other.to_seconds())
    }
}

#[cfg_attr(tarpaulin, skip)]
impl Ord for TimeTuple {
    fn cmp(&self, other: &TimeTuple) -> Ordering {
        self.to_seconds().cmp(&other.to_seconds())
    }
}

impl Add for TimeTuple {
    type Output = TimeTuple;
    fn add(self, other: TimeTuple) -> TimeTuple {
        TimeTuple::new(
            i32::from(self.h + other.h),
            i32::from(self.m + other.m),
            i32::from(self.s + other.s),
        )
    }
}

impl AddAssign for TimeTuple {
    fn add_assign(&mut self, other: TimeTuple) {
        *self = TimeTuple::new(
            i32::from(self.h + other.h),
            i32::from(self.m + other.m),
            i32::from(self.s + other.s),
        );
    }
}

impl Sub for TimeTuple {
    type Output = TimeTuple;
    fn sub(self, other: TimeTuple) -> TimeTuple {
        TimeTuple::new(
            i32::from(self.h - other.h),
            i32::from(self.m - other.m),
            i32::from(self.s - other.s),
        )
    }
}

impl SubAssign for TimeTuple {
    fn sub_assign(&mut self, other: TimeTuple) {
        *self = TimeTuple::new(
            i32::from(self.h - other.h),
            i32::from(self.m - other.m),
            i32::from(self.s - other.s),
        );
    }
}

/// A wrapper for a duration.
///
/// Does not count in days, but can have hours >24 (up to `u32::MAX`)
///
/// Precise to second-level.
///
/// Cannot be negative.
#[derive(PartialEq, Eq, Debug, Copy, Clone)]
pub struct Duration {
    h: u32,
    m: u8,
    s: u8,
}

impl Duration {
    /// Produces a new Duration.
    ///
    /// The value is calculated from total number of seconds so a duration
    /// with a minute value of 90 would add an hour to the resulting tuple
    /// and set the minutes to 30, for example.
    pub fn new(h: u32, m: u32, s: u32) -> Duration {
        let total_seconds: u64 = u64::from(s) + 60 * u64::from(m) + 3600 * u64::from(h);
        Duration::from_seconds(total_seconds)
    }

    /// Same as `Duration::new()` but takes the total number of seconds
    /// as its argument and calculates the hours, minutes, and seconds
    /// from that.
    pub fn from_seconds(mut total_seconds: u64) -> Duration {
        let h = total_seconds / 3600;
        total_seconds -= h * 3600;
        let m = total_seconds / 60;
        total_seconds -= m * 60;
        Duration {
            h: h as u32,
            m: m as u8,
            s: total_seconds as u8,
        }
    }

    /// Calculates the `Duration` between two `DateTimeTuple`s.
    pub fn between(dt1: DateTimeTuple, dt2: DateTimeTuple) -> Duration {
        if dt1 == dt2 {
            return Duration { h: 0, m: 0, s: 0 };
        }
        let smaller = if dt1 < dt2 { dt1 } else { dt2 };
        let greater = if dt1 < dt2 { dt2 } else { dt1 };
        let days_between = greater.get_date().to_days() - smaller.get_date().to_days();
        if days_between == 0 {
            Duration::from(greater.get_time()) - Duration::from(smaller.get_time())
        } else {
            let time_between = Duration::from(greater.get_time()) + Duration::new(24, 0, 0)
                - Duration::from(smaller.get_time());
            time_between + Duration::new(24 * (days_between - 1), 0, 0)
        }
    }

    pub fn get_hours(self) -> u32 {
        self.h
    }

    pub fn get_minutes(self) -> u8 {
        self.m
    }

    pub fn get_seconds(self) -> u8 {
        self.s
    }

    /// Produces a string such as 8:30 for 8 hours and 30 minutes.
    ///
    /// Hours field will expand as necessary; 150:30 is a possible result.
    ///
    /// Ignores seconds.
    pub fn to_hhmm_string(self) -> String {
        format!("{}:{:02}", self.h, self.m)
    }

    /// Produces a string such as 8:30 for 8 hours and 30 minutes.
    ///
    /// Hours field will expand as necessary; 150:30 is a possible result.
    ///
    /// Ignores seconds.
    #[deprecated(since = "2.1.0", note = "Replace with to_hhmm_string()")]
    pub fn to_hours_and_minutes_string(self) -> String {
        format!("{}:{:02}", self.h, self.m)
    }

    /// Gets the total number of seconds in the Duration.
    pub fn to_seconds(self) -> u64 {
        3600 * u64::from(self.h) + 60 * u64::from(self.m) + u64::from(self.s)
    }

    /// Gets the total number of minutes in the Duration, ignoring seconds.
    pub fn to_minutes(self) -> u32 {
        60 * self.h + u32::from(self.m)
    }

    /// Adds a number of seconds to the Duration,
    /// wrapping the same way `Duration::new()` does.
    pub fn add_seconds(&mut self, seconds: u32) {
        let new_seconds = u32::from(self.s) + seconds;
        *self = Duration::new(self.h, u32::from(self.m), new_seconds);
    }

    /// Subtracts a number of seconds from the Duration,
    /// wrapping the same way `Duration::new()` does.
    pub fn subtract_seconds(&mut self, seconds: u32) {
        *self = Duration::from_seconds(self.to_seconds() - u64::from(seconds));
    }

    /// Adds a number of minutes to the Duration,
    /// wrapping the same way `Duration::new()` does.
    pub fn add_minutes(&mut self, minutes: u32) {
        let new_minutes = u32::from(self.m) + minutes;
        *self = Duration::new(self.h, new_minutes, u32::from(self.s));
    }

    /// Subtracts a number of minutes from the Duration,
    /// wrapping the same way `Duration::new()` does.
    pub fn subtract_minutes(&mut self, minutes: u32) {
        *self = Duration::from_seconds(self.to_seconds() - u64::from(minutes) * 60);
    }

    /// Adds a number of hours to the Duration,
    /// wrapping the same way `Duration::new()` does.
    pub fn add_hours(&mut self, hours: u32) {
        let new_hours = self.h + hours;
        *self = Duration::new(new_hours, u32::from(self.m), u32::from(self.s));
    }

    /// Subtracts a number of hours from the Duration,
    /// wrapping the same way `Duration::new()` does.
    pub fn subtract_hours(&mut self, hours: u32) {
        let new_hours = self.h - hours;
        *self = Duration::new(new_hours, u32::from(self.m), u32::from(self.s));
    }
}

impl fmt::Display for Duration {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}:{:02}:{:02}", self.h, self.m, self.s)
    }
}

impl FromStr for Duration {
    type Err = String;

    fn from_str(s: &str) -> Result<Duration, Self::Err> {
        lazy_static! {
            static ref VALID_FORMAT: Regex = Regex::new(r"^\d+:\d{2}:\d{2}$").unwrap();
        }
        if !VALID_FORMAT.is_match(s) {
            Err(format!(
                "Invalid str formatting of Duration: {}\nExpects a string formatted like 8:30:05",
                s
            ))
        } else {
            let mut parts = s.split(':');
            Ok(Duration::new(
                u32::from_str(parts.next().unwrap()).unwrap(),
                u32::from_str(parts.next().unwrap()).unwrap(),
                u32::from_str(parts.next().unwrap()).unwrap(),
            ))
        }
    }
}

impl PartialOrd for Duration {
    fn partial_cmp(&self, other: &Duration) -> Option<Ordering> {
        self.to_seconds().partial_cmp(&other.to_seconds())
    }
}

#[cfg_attr(tarpaulin, skip)]
impl Ord for Duration {
    fn cmp(&self, other: &Duration) -> Ordering {
        self.to_seconds().cmp(&other.to_seconds())
    }
}

impl Add for Duration {
    type Output = Duration;
    fn add(self, other: Duration) -> Duration {
        Duration::new(
            self.h + other.h,
            u32::from(self.m + other.m),
            u32::from(self.s + other.s),
        )
    }
}

impl AddAssign for Duration {
    fn add_assign(&mut self, other: Duration) {
        *self = Duration::new(
            self.h + other.h,
            u32::from(self.m + other.m),
            u32::from(self.s + other.s),
        );
    }
}

impl Sub for Duration {
    type Output = Duration;
    fn sub(self, other: Duration) -> Duration {
        Duration::new(
            self.h - other.h,
            u32::from(self.m - other.m),
            u32::from(self.s - other.s),
        )
    }
}

impl SubAssign for Duration {
    fn sub_assign(&mut self, other: Duration) {
        *self = Duration::new(
            self.h - other.h,
            u32::from(self.m - other.m),
            u32::from(self.s - other.s),
        );
    }
}

impl From<TimeTuple> for Duration {
    fn from(time: TimeTuple) -> Self {
        Duration::from_seconds(u64::from(time.to_seconds()))
    }
}

#[cfg(test)]
mod tests {

    use super::TimeTuple;

    #[test]
    fn test_no_seconds() {
        let tuple = TimeTuple::new(5, 30, 0);
        assert_eq!(5, tuple.h);
        assert_eq!(30, tuple.m);
        assert_eq!(0, tuple.s);
    }

    #[test]
    fn test_no_overlap() {
        let tuple = TimeTuple::new(5, 30, 30);
        assert_eq!(5, tuple.h);
        assert_eq!(30, tuple.m);
        assert_eq!(30, tuple.s);
    }

    #[test]
    fn test_second_overlap() {
        let tuple = TimeTuple::new(6, 30, 90);
        assert_eq!(6, tuple.h);
        assert_eq!(31, tuple.m);
        assert_eq!(30, tuple.s);
    }

    #[test]
    fn test_minute_overlap() {
        let tuple = TimeTuple::new(6, 90, 30);
        assert_eq!(7, tuple.h);
        assert_eq!(30, tuple.m);
        assert_eq!(30, tuple.s);
    }

    #[test]
    fn test_hour_overlap() {
        let tuple = TimeTuple::new(25, 30, 30);
        assert_eq!(1, tuple.h);
        assert_eq!(30, tuple.m);
        assert_eq!(30, tuple.s);
    }

    #[test]
    fn test_all_overlap() {
        let tuple = TimeTuple::new(25, 90, 90);
        assert_eq!(2, tuple.h);
        assert_eq!(31, tuple.m);
        assert_eq!(30, tuple.s);
    }

    #[test]
    fn test_minutes_to_hours_overlap() {
        let tuple = TimeTuple::new(18, 420, 0);
        assert_eq!(1, tuple.h);
        assert_eq!(0, tuple.m);
        assert_eq!(0, tuple.s);
    }

    #[test]
    fn test_negative_seconds() {
        let tuple = TimeTuple::new(6, 30, -60);
        assert_eq!(6, tuple.h);
        assert_eq!(29, tuple.m);
        assert_eq!(0, tuple.s);
    }

    #[test]
    fn test_all_negative_overlaps() {
        let tuple = TimeTuple::new(-3, -116, -301);
        assert_eq!(18, tuple.h);
        assert_eq!(58, tuple.m);
        assert_eq!(59, tuple.s);
    }
}
